### DISCUSSION TITLE
send byte array to SerialWrite

### DISCUSSION THREAD

**Discussion by KathaSchmack** [bonsai-rx/bonsai#1724] [Q&A] [answered]
Hi,

I would like to stream the position of the Bpod Rotary Encoder Module to Bonsai. I am connecting the Rotary Encoder Module via its own serial port to Bonsai using CreateSerialPort. To start streaming, I would like to send these bytes "(ord('S'), 1)" [Python] to the serial port via SerialWrite. How can I create a byte array that contains both the unicode for S and an integer for 1? 

Many thanks,
Katharina

P.S.: here is the related discussion on the Bpod Forum https://sanworks.io/forums/d/3944-streaming-of-rotary-encoder-position-to-bonsai

**Comment by VegaCaroMiguel** [comment]
Hi @KathaSchmack,

In order to open a node buffer on a serial port I think default __init__ python bytes() function might work:
For the position encoding you would have two options:
1. bytes() and bytearray():
       The difference between bytes() and bytearray() is that bytes() returns an object that cannot be modified, and bytearray() returns an object 
       that can be modified.
       -bytes(x, encoding, error) vs bytearray(x, encoding, error)

You need a **constant stream** of data regarding to the position I guess, so: 
> Byte array creation with "S" to integer plain val and 1 as encoder 
byte_array = bytes([ord('S'), 1])

> Print the byte array to check the node endpoint 
print(byte_array)

Now for sending, if you're using Python inbuilt serial lib then: 

import serial
> Gate the port 
stream_serial =  serial.Serial('WRITE_THE_SERIALPORTHERE', WRITE_THE_BAUDRATEHERE) 

> Send the data stream 
stream_serial.write(byte_array)

> Close stream serial
stream_serial.close()

If you are using **ROS**, it's another history, Id be glad to help you ! 
Have a nice one !

**Reply by KathaSchmack** [reply]
Thank you @VegaCaroMiguel, that sounds like a good solution for working from Python. 

However, I would like to do that from Bonsai-Rx, something like this but it is currently not working.

Thank you! 

![image](https://github.com/bonsai-rx/bonsai/assets/95446942/1e0db2c7-2df3-415a-972e-9b72fd968eb0)
[RotaryEncoder.bonsai.zip](https://github.com/bonsai-rx/bonsai/files/14811528/RotaryEncoder.bonsai.zip)

**Reply by bruno-f-cruz** [reply]
Heya,
Here are a few ways of manipulating the data types I think you might need. As I side note, I think you would have a much easier time using the SerialWriteLine operator than working with the byte[] directly :)

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/5d01f570-1803-4569-8441-e3672a503af5)


<details>
    <summary>Workflow (Copy+Paste me)</summary>
    
```xml

<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.8.1"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:port="clr-namespace:Bonsai.IO.Ports;assembly=Bonsai.System"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns:scr="clr-namespace:Bonsai.Scripting.Expressions;assembly=Bonsai.Scripting.Expressions"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="port:CreateSerialPort">
          <port:BaudRate>9600</port:BaudRate>
          <port:NewLine>\r\n</port:NewLine>
          <port:Parity>None</port:Parity>
          <port:ParityReplace>63</port:ParityReplace>
          <port:DataBits>8</port:DataBits>
          <port:StopBits>One</port:StopBits>
          <port:Handshake>None</port:Handshake>
          <port:DiscardNull>false</port:DiscardNull>
          <port:DtrEnable>false</port:DtrEnable>
          <port:RtsEnable>false</port:RtsEnable>
          <port:ReadBufferSize>4096</port:ReadBufferSize>
          <port:WriteBufferSize>2048</port:WriteBufferSize>
          <port:ReceivedBytesThreshold>1</port:ReceivedBytesThreshold>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="StringProperty">
          <Value>FooCommand</Value>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="port:SerialWriteLine" />
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="ByteProperty">
          <Value>0</Value>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Take">
          <rx:Count>1</rx:Count>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="ByteProperty">
          <Value>0</Value>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Take">
          <rx:Count>1</rx:Count>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Merge" />
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:ToArray" />
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="port:SerialWrite" />
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="StringProperty">
          <Value>MyCommandFoo</Value>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Merge" />
      </Expression>
      <Expression xsi:type="scr:ExpressionTransform">
        <scr:Expression>Convert.ToByte(it)</scr:Expression>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:ToArray" />
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="port:SerialWrite" />
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="1" To="2" Label="Source1" />
      <Edge From="3" To="4" Label="Source1" />
      <Edge From="4" To="7" Label="Source1" />
      <Edge From="5" To="6" Label="Source1" />
      <Edge From="6" To="7" Label="Source2" />
      <Edge From="7" To="8" Label="Source1" />
      <Edge From="8" To="9" Label="Source1" />
      <Edge From="10" To="11" Label="Source1" />
      <Edge From="11" To="12" Label="Source1" />
      <Edge From="12" To="13" Label="Source1" />
      <Edge From="13" To="14" Label="Source1" />
    </Edges>
  </Workflow>
</WorkflowBuilder>

```

</details>

**Comment by KathaSchmack** [comment]
Amazing, thank you so much @bruno-f-cruz ! I can't wait to try this out as soon as I am back to my rotary encoder, especially the SerialWriteLine seems like a very convenient way of sending commands to the SerialPort!

**Comment by KathaSchmack** [comment]
That worked, we can now stream from the Rotary Encoder, thank you @bruno-f-cruz !

**Reply by KathaSchmack** [reply]
Now, we also would like to read from the Rotary Encoder. Here is how the Rotary Encoder sends the position
![image](https://github.com/bonsai-rx/bonsai/assets/95446942/f82d61be-0d2f-41fd-9f4d-ab7943ce129e)
https://sanworks.github.io/Bpod_Wiki/serial-interfaces/rotary-encoder-module-serial-interface/

I am able to receive bytes from the Rotary Encoder, yeah! However, I cannot figure out the right way of transforming them into the integer indicating the rotary position in tics (I understand that this should be 1 to 1024). 
![image](https://github.com/bonsai-rx/bonsai/assets/95446942/2821899a-cd2c-4503-a591-560dbd0b34bc)
[RotaryEncoder.zip](https://github.com/bonsai-rx/bonsai/files/15026181/RotaryEncoder.zip)

Many thanks,
Katharina

**Reply by bruno-f-cruz** [reply]
I am not familiar with this bespoke protocol. This kind of communication can be made significantly simpler from the client side if:

1. Do the message from the device share the same length?
2. Is there a CR or NL character that can be used to split messages?

If the answer to the two previous questions is no, I would probably just code something directly in csharp or python to parse the protocol. Doing the whole parsing logic using visual operators seems insane :P

**Comment by KathaSchmack** [comment]
Thanks!

I believe that 
1. The message is always 7 bytes of which I want the 2nd and 3rd bytes as a 16-bit integer
2. There are nor CR/NL characters

Coding a parser in csharp is a great idea, I will install all the packages and give it try (although I was just starting to have fun with the visual logic;)

**Reply by KathaSchmack** [reply, marked as answer]
I got this to work with SerialWriteLine and a csharp script, thank you @bruno-f-cruz for the help and suggestions! Here is the full workflow for using the Bpod Rotary Encoder Module in Bonsai, in case anyone coming across this post might find it useful. 
[RotaryEncoder.zip](https://github.com/bonsai-rx/bonsai/files/15039979/RotaryEncoder.zip)
